### PR TITLE
#17843 Fallback to Tomcat provider if the datasource initialization fails

### PR DESCRIPTION
**Pull request by nollymar**
This logic was implemented to use context.xml as a fallback in case the datasource initialization fails using another provider

### CODE REVIEW

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java [hunk 1]
```diff
@@ -99,30 +106,84 @@ public void testGet(final String testCase)
         Mockito.verify(dbStrategy, Mockito.times(testCase.equals("DBProperties")? 1: 0)).apply();
         Mockito.verify(systemEnvStrategy, Mockito.times(testCase.equals("SystemEnv")? 1: 0)).apply();
         Mockito.verify(dockerSecretStrategy, Mockito.times(testCase.equals("DockerSecret")? 1: 0)).apply();
-        Mockito.verify(tomcatDataSourceStrategy, Mockito.times(testCase.equals("Tomcat")? 1: 0)).apply();
+        Mockito.verify(tomcatDataSourceStrategy, Mockito.times(
+                (testCase.equals("Tomcat") | testCase.equals("CustomProvider")) ? 1 : 0)).apply();
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/main/java/com/dotmarketing/db/DataSourceStrategyProvider.java [hunk 2]
```diff
@@ -59,38 +59,47 @@ public DataSource get()
...
+                DotDataSourceStrategy customStrategy = ((Class<DotDataSourceStrategy>) Class
+                        .forName(providerClassName)).newInstance();
+                defaultDataSource = customStrategy.apply();
+
                 Logger.info(DbConnectionFactory.class,
-                        "Datasource loaded from Docker Secret");
+                        "Datasource loaded using custom class " + providerClassName);
             }
 
+        } catch(Exception e) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/db/DataSourceStrategyProvider.java]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java (hunk 1 above)
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java [hunk 3]
```diff
@@ -27,18 +27,21 @@
                 "DBProperties",
                 "SystemEnv",
                 "DockerSecret",
-                "Tomcat"
+                "Tomcat",
+                "CustomProvider"
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [The String literal "CustomProvider" appears 4 times in this file; the first occurrence is on line 31](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java [hunk 4]
```diff
@@ -99,30 +106,84 @@ public void testGet(final String testCase)
...
+            Mockito.when(properties.getVariable("connection_db_base_url")).thenReturn(null);
+        }
+
+        if (testCase.equals("DockerSecret")) {
+            Mockito.when(dockerSecretStrategy.dockerSecretPathExists()).thenReturn(true);
+        } else {
+            Mockito.when(dockerSecretStrategy.dockerSecretPathExists()).thenReturn(false);
+        }
+
+        if (testCase.equals("CustomProvider")){
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/main/java/com/dotmarketing/db/DataSourceStrategyProvider.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/db/DataSourceStrategyProvider.java]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java [hunk 5]
```diff
@@ -99,30 +106,85 @@ public void testGet(final String testCase)
         Mockito.verify(dbStrategy, Mockito.times(testCase.equals("DBProperties")? 1: 0)).apply();
         Mockito.verify(systemEnvStrategy, Mockito.times(testCase.equals("SystemEnv")? 1: 0)).apply();
         Mockito.verify(dockerSecretStrategy, Mockito.times(testCase.equals("DockerSecret")? 1: 0)).apply();
-        Mockito.verify(tomcatDataSourceStrategy, Mockito.times(testCase.equals("Tomcat")? 1: 0)).apply();
+        Mockito.verify(tomcatDataSourceStrategy, Mockito.times(
+                (testCase.equals("Tomcat") | testCase.equals("CustomProvider")) ? 1 : 0)).apply();
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/main/java/com/dotmarketing/db/DBPropertiesDataSourceStrategy.java [hunk 6]
```diff
@@ -63,7 +64,7 @@ public DataSource apply() {
                 throw new FileNotFoundException("DB properties file not found");
             }
 
-            properties.load(new FileInputStream(propertiesFile));
+            properties.load(new FileInputStream(getPropertiesFile()));
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/db/DBPropertiesDataSourceStrategy.java]: Codacy Issue found: [Avoid instantiating FileInputStream, FileOutputStream, FileReader, or FileWriter](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [The String literal "CustomProvider" appears 4 times in this file; the first occurrence is on line 31](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java [hunk 7]
```diff
@@ -99,30 +106,85 @@ public void testGet(final String testCase)
...
+            Mockito.when(properties.getVariable("connection_db_base_url")).thenReturn(null);
+        }
+
+        if (testCase.equals("DockerSecret")) {
+            Mockito.when(dockerSecretStrategy.dockerSecretPathExists()).thenReturn(true);
+        } else {
+            Mockito.when(dockerSecretStrategy.dockerSecretPathExists()).thenReturn(false);
+        }
+
+        if (testCase.equals("CustomProvider")){
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java (hunk 5 above)
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [The String literal "CustomProvider" appears 4 times in this file; the first occurrence is on line 31](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/main/java/com/dotmarketing/db/DataSourceStrategyProvider.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/db/DataSourceStrategyProvider.java]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/main/java/com/dotmarketing/db/DBPropertiesDataSourceStrategy.java (hunk 6 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/db/DBPropertiesDataSourceStrategy.java]: Codacy Issue found: [Avoid instantiating FileInputStream, FileOutputStream, FileReader, or FileWriter](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java (hunk 7 above)
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java (hunk 5 above)
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [The String literal "CustomProvider" appears 4 times in this file; the first occurrence is on line 31](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java (hunk 7 above)
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/main/java/com/dotmarketing/db/DataSourceStrategyProvider.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/db/DataSourceStrategyProvider.java]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/main/java/com/dotmarketing/db/DBPropertiesDataSourceStrategy.java (hunk 6 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/db/DBPropertiesDataSourceStrategy.java]: Codacy Issue found: [Avoid instantiating FileInputStream, FileOutputStream, FileReader, or FileWriter](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

File: dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java (hunk 5 above)
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotmarketing/db/DataSourceStrategyProviderTest.java]: Codacy Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5092420)

**Review by freddyucv — APPROVED** [review]

**Review by dsilvam — APPROVED** [review]

**Pull request merged**